Zinc 157: fix handling of anonymous java classes

romanowski opened the pull request:
Fix for #157 and more including:
- fix enums handling
- fix problem with 1 ms file changes
- create external logging mechanism for e.g. IDE integration

For certain reasons this time I have to paste the below disclaimer:
THIS PROGRAM IS SUBJECT TO THE TERMS OF THE BSD 3-CLAUSE LICENSE.

THE FOLLOWING DISCLAIMER APPLIES TO ALL SOFTWARE CODE AND OTHER MATERIALS CONTRIBUTED IN CONNECTION WITH THIS SOFTWARE:
THIS SOFTWARE IS LICENSED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS “AS IS” AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE AND ANY WARRANTY OF NON-INFRINGEMENT, ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. THIS SOFTWARE MAY BE REDISTRIBUTED TO OTHERS ONLY BY EFFECTIVELY USING THIS OR
ANOTHER EQUIVALENT DISCLAIMER IN ADDITION TO ANY OTHER REQUIRED LICENSE TERMS.
ONLY THE SOFTWARE CODE AND OTHER MATERIALS CONTRIBUTED IN CONNECTION WITH THIS SOFTWARE, IF ANY, THAT ARE ATTACHED TO (OR OTHERWISE ACCOMPANY) THIS SUBMISSION (AND ORDINARY
COURSE CONTRIBUTIONS OF FUTURES PATCHES THERETO) ARE TO BE CONSIDERED A CONTRIBUTION. NO OTHER SOFTWARE CODE OR MATERIALS ARE A CONTRIBUTION.

romanowski committed:
Fix enum handling from java code

romanowski committed:
Ingnore changes smaller them 1ms due to problem on windows

eed3si9n commented on zinc/src/main/scala/sbt/internal/inc/IncrementalCompilerImpl.scala:
```diff
@@ -8,17 +8,18 @@ import xsbti.compile.{ CompileOrder, GlobalsCache, IncOptions, MiniSetup, Compil
 //     by this class.
 
-class IncrementalCompilerImpl extends IncrementalCompiler {
+class IncrementalCompilerImpl(externalLookup: Option[ExternalLookup] = None) extends IncrementalCompiler {
```
I think ExternalLookup should be added to `xsbti.compile.Setup` like other setup related things. See my half done pull request https://github.com/sbt/zinc/pull/131.

lightbend-cla-validator commented:
Hi @romanowski,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Typesafe Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

romanowski committed:
In case of lack of cannonical name point to enclosing class if exsist.

Fix java compilation problems (recompiles) for annonymous classes.

romanowski committed:
Create extrenal hooks for zinc

Users can now configure external ClassfileManager (that will work alongside zinc's internal one) and ExternalLookup that can override checks for initially modified sources, products or binaries.

lightbend-cla-validator commented:
Hi @romanowski,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Typesafe Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

eed3si9n commented:
LGTM
